### PR TITLE
scrape: provide a fallback format

### LINKED ISSUE #13214: scrape: Bail out of invalid scrape format rather than silently falling back to the classic text format
### Proposal

Back in the old days, it was a reasonable assumption that a target that fails to properly negotiate the scrape format is probably an ad-hoc implementation that most likely tries to produce the (classic) text format. Therefore, it was a good idea to fall back to parsing the classic text format whenever anything went wrong.

However, this assumption doesn't hold anymore. It might even be dangerous. That's mostly because now there is OpenMetrics, which looks very similar to the classic Prometheus text format but is not the same. With the current fallback behavior, Prometheus might end up in a situation where it parses OpenMetrics as the classic text format, which - in the worst case - works without error but yielding wrong results (most relevant when it comes to timestamps).

A recent issue caused by this is #13172. In short, the following happened: With native histograms enabled, Prometheus tries to negotiate protobuf. However, [KSM](https://github.com/kubernetes/kube-state-metrics) wasn't prepared for that and responded with some invalid content type (details to be investigated), which caused Prometheus to parse the KSM scrape as classic text format, while KSM was serving OpenMetrics. Luckily, the metrics exposed by KSM use features of OM that won't parse as the classic text format, so Prometheus error'd out rather than silently ingesting wrong data. However, the error message is still quite confusing (see #13272) and it would have been much clearer if Prometheus had just reported that KSM wasn't negotiating a valid content type.

So the proposal here is to simply not do a fallback but report an error in case of invalid content negotiation.
(Relevant code: https://github.com/prometheus/prometheus/blob/b8bcaef14da44eb384e8a07655da35614dc97467/scrape/scrape.go#L1404-L1410 https://github.com/prometheus/prometheus/blob/b8bcaef14da44eb384e8a07655da35614dc97467/model/textparse/interface.go#L81-L91 )

This is a breaking change as currently there might be a lot of targets that aren't doing proper content negotiation but "just work". This change would break those use cases, but with the danger of ingesting wrong data (as explained above), we should make use of the 3.0.0 release to deliberately expose those broken scrapers.

An open question is if we want to retain the simplicity of doing trivial text-format expositions without _any_ kind of content negotiation, i.e. the [case where the content type ends up empty](https://github.com/prometheus/prometheus/blob/b8bcaef14da44eb384e8a07655da35614dc97467/model/textparse/interface.go#L84-L86). OpenMetrics has created a dilemma here. The subtle differences between OM and the classic text format create ambiguities (see above). We could retain the feature of using one specific format in case the target doesn't do any content negotiation. But which format would that be? We have only bad choices here:
1. If we stuck with the classic text format, the existing use cases won't break, but we would extend the life of the deprecated classic text format forever.
2. If we fell back to OpenMetrics, we would run into the danger of ingesting old targets in the wrong way (most relevant concern: timestamps).
3. We could only accept a "safe common subset" of OM and the classic format in case of missing content negotiation. This would only break a few use cases, but it would be a weird and complicated contraption to cater for a supposedly "trivial" use case.

_(Edited for clarity and correctness. Previously I assumed, Prometheus would also fall back on parsing problem when one content type was negotiated but another one was served, but it doesn't.)_

### PR DESCRIPTION
**Pull request by alexgreenbank**
Fixes #13214

Provides a per scrape target setting for a fallback format for scraping that is used if the received `Content-Type` does not match an expected value, or is missing/blank.

This is a breaking change as some scrapes that worked with Prometheus 2.x may now require the `fallback_scrape_protocol` being set in order to force the scrape.

TODO
- [x] Documentation (changelog)
- [x] Documentation (breaking change / Prometheus 3.0 upgrade Docs): https://github.com/prometheus/prometheus/pull/15099 <-- fixed link

### CODE REVIEW

**Comment by beorn7** [comment]
:thinking: I actually had a quite different idea of how to approach this.

- I think the default behavior in v3 should be to fail the scrape if anything goes wrong during content negotiation, including not returning a content-type at all. That's the only safe behavior. (After all, #13214 was triggered by the insight that there is no safe default anymore, as a PromParser could parse OpenMetrics without error (or vice versa) but yielding wrong results (most notably timestamps).)
- If a fallback format is specified, the scrape should proceed, but the fallback should be logged. (Maybe we could consider to not log anything if the returned content-type is empty, to mirror the v2 behavior.)
- I would prefer to not use a command line flag, but a setting per-scrape-config (next to the existing `scrape_protocols`, e.g. `fallback_scrape_protocol`).

**Comment by beorn7** [comment]
Maybe we should get some more opinions on this to find out which way to go. @bwplotka you had some thoughts on the original issue, but everyone's feedback welcome.

**Comment by alexgreenbank** [comment]
> 🤔 I actually had a quite different idea of how to approach this.

OK, that all makes sense, I'll knock up an alternative approach that is per scrape target (I did play with something like that originally) and takes into account these points.

**Comment by alexgreenbank** [comment]
For the protocol names, are the existing mime types appropriate for use as the values for `fallback_scrape_protocol`?
* `text/plain` for prometheus
* `application/openmetrics-text` for openmetrics
* `application/vnd.google.protobuf` for protobuf
?

**Comment by bwplotka** [comment]
I agree with @beorn7 directions here: https://github.com/prometheus/prometheus/pull/15136#issuecomment-2403290709 - this is what I would have in mind too.

> For the protocol names, are the existing mime types appropriate for use as the values for fallback_scrape_protocol?
> text/plain for prometheus
> application/openmetrics-text for openmetrics
> application/vnd.google.protobuf for protobuf

What do you mean, why not reusing existing protocol enums in [`scrape_protocols`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config:~:text=%23%20The%20protocols%20to%20negotiate%20during%20a%20scrape%20with%20the%20client.%0A%23%20Supported%20values%20(case%20sensitive)%3A%20PrometheusProto%2C%20OpenMetricsText0.0.1%2C%0A%23%20OpenMetricsText1.0.0%2C%20PrometheusText0.0.4.)

**Comment by alexgreenbank** [comment]
> What do you mean, why not reusing existing protocol enums in [`scrape_protocols`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config:~:text=%23%20The%20protocols%20to%20negotiate%20during%20a%20scrape%20with%20the%20client.%0A%23%20Supported%20values%20(case%20sensitive)%3A%20PrometheusProto%2C%20OpenMetricsText0.0.1%2C%0A%23%20OpenMetricsText1.0.0%2C%20PrometheusText0.0.4.)

Ah ok. I'd been looking at the various existing scrape tests which focused on dealing with the raw `Content-Type` values. (This includes `text/plain; version=1.0.0; escaping=allow-utf-8` - but there hasn't been a corresponding change in the list of valid `scrape_protocols` in order to cover this. I should probably add this whilst I'm here.)

So, for my brain:
* `scrape_protocols` helps define the `Accept` header Prometheus sends when attempting the scrape. It's not used at all when processing the scrape results.
* `fallback_scrape_protocol` (using one of the same enums) will help determine the behaviour of Prometheus when parsing the scrape response if the returned `Content-Type` is missing, unparseable (as a `mimeType`) or not an expected value.

v2 does the following:
* No or blank `Content-Type` -> PromParser, no error logged
* Unparseable `Content-Type` -> PromParser, error from `mime.ParseMediaType()` logged
* `mime.ParseMediaType(Content-Type)` == `application/openmetrics-text` -> OpenMetricsParser, no error logged
* `mime.ParseMediaType(Content-Type)` == `application/vnd.google.protobuf` -> ProtobufParser, no error logged
* Anything else -> PromParser, no error logged

v3 should do:
* No `Content-Type`, no `fallback_scrape_protocol` -> fail the scrape, log error about failed scrape
* No `Content-Type`, `fallback_scrape_protocol` specified -> use fallback parser, log warning about use of fallback (**)
* Unparseable `Content-Type`, no `fallback_scrape_protocol` -> fail the scrape, error logged about failed scrape and invalid/unparseable `Content-Type`
* Unparseable `Content-Type`, `fallback_scrape_protocol` specified -> use fallback parser, log warning about use of fallback and invalid/unparseable `Content-Type`
* `mime.ParseMediaType(Content-Type)` == `application/openmetrics-text` -> OpenMetricsParser, no error logged
* `mime.ParseMediaType(Content-Type)` == `application/vnd.google.protobuf` -> ProtobufParser, no error logged
* `mime.ParseMediaType(Content-Type)` == `text/plain` -> ProtobufParser, no error logged
* Anything else (e.g. some unsupported `Content-Type`), no `fallback_scrape_protocol` -> fail the scrape, error logged about failed scrape and invalid/unparseable `Content-Type`
* Anything else, `fallback_scrape_protocol` specified -> use fallback parser, log warning about use of fallback and invalid/unparseable `Content-Type`

The second one (**) is the one where we could not log anything to mimic v2 behaviour.

**Comment by beorn7** [comment]
> mime.ParseMediaType(Content-Type) == text/plain -> ProtobufParser, no error logged

I guess this should be PromParser instead of ProtobufParser.

Otherwise, sounds spot on.

About the (**) case: I wouldn't worry too much. The logging in v2 is just debug level. I don't think anyone will complain about this changing slightly.

The way more impactful change is that we might fail scrapes now that succeeded before, and you need a config change to make them succeed again. And that is the reason why it would be great to get this into the v3 release.

**Comment by alexgreenbank** [comment]
> I guess this should be PromParser instead of ProtobufParser.

Indeed.

OK, Revamped to be per scrape target.

Will go look for Jan's migration to 3.0 guide as it would be useful to add some comments in there.

I did add a new `ScrapeProtocol` for 1.0.0 which is UTF-8 aware: `PrometheusText1.0.0`. I assume we want to have this before `PrometheusText0.0.4` in the ordering so that a well implemented client will see we want 1.0.0 with higher priority than 0.0.4.

**Comment by alexgreenbank** [comment]
Added a review comment with suggested text to Jan's 3.0 migration PR (https://github.com/prometheus/docs/pull/2530)

File: model/textparse/interface.go [hunk 1]
```diff
@@ -78,28 +79,66 @@ type Parser interface {
...
+	// or we have to error
+	switch mediaType {
+	case "application/openmetrics-text", "application/vnd.google.protobuf", "text/plain":
+		return mediaType, nil
+	}
+	// We're here because we have no regonised mediaType
+	if fallbackType == "" {
+		return "", errors.New("Scraper sending unrecognisable Content-Type and no fallback_scrape_protocol specified for target")
 	}
+	return fallbackType, errors.New("Content-Type not recognised mediaType, using fallback_scrape_protocol for target")
```
Comment (beorn7) [inline model/textparse/interface.go:116]: Let's include the invalid content-type string in the error message.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks good in general. 

The only substantial change request here is that I think we should also log if the content type is blank.

**Comment by beorn7** [comment]
"log an error" sounds very harsh, it's just a debug log, so off by default. I really think we should log any use of the fallback at this level.

File: model/textparse/interface.go [hunk 2]
```diff
@@ -78,28 +79,66 @@ type Parser interface {
...
+// first and falling back to fallbackType if necessary.
+// As fallbackType comes from the parse `fallback_scrape_protocol` config option it
+// is assumed to be a valid/supported mediaType.
+func newHelper(contentType, fallbackType string) (string, error) {
 	if contentType == "" {
-		return NewPromParser(b, st), nil
+		if fallbackType == "" {
+			return "", errors.New("Non-compliant scraper sending blank Content-Type and no fallback_scrape_protocol specified for target")
+		}
+		// We don't set the error here so that we're kind and mirror v2 behaviour
```
[the remark this answers is not in the document]
Comment (alexgreenbank) [reply model/textparse/interface.go]: OK, added a similar error message and updated the tests.

File: model/textparse/interface.go [hunk 3]
```diff
@@ -78,28 +79,66 @@ type Parser interface {
...
+		return fallbackType, nil
 	}
 
+	// We have a contentType, parse it
```
[the remark this answers is not in the document]
Comment (alexgreenbank) [reply model/textparse/interface.go]: Added and fixed other comments I had added. Getting used to these stylistic conventions...

[the remark this answers is not in the document]
Comment (beorn7) [reply model/textparse/interface.go]: The "end with a period" is essentially an extension of the [official style guide for doc comments](https://go.dev/wiki/CodeReviewComments#comment-sentences) to code comments. One day I'll write a Prometheus Go styleguide just listing the few additional quirks we have added over time.

File: model/textparse/interface.go (hunk 1 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Done (and to the corresponding error above for no fallback) and fixed the associated tests.

File: model/textparse/interface.go [hunk 4]
```diff
@@ -78,28 +80,65 @@ type Parser interface {
...
-func New(b []byte, contentType string, parseClassicHistograms, skipOMCTSeries bool, st *labels.SymbolTable) (Parser, error) {
+// extractMediaType returns the mediaType of a required parser. It tries first to
+// extract a valid and supported mediaType from contentType. If that fails,
+// the provided fallbackType (possibly an empty string) is returned, together with
+// an error. fallbackType is used as-is without further validation.
+func extractMediaType(contentType, fallbackType string) (string, error) {
 	if contentType == "" {
-		return NewPromParser(b, st), nil
+		if fallbackType == "" {
+			return "", errors.New("Non-compliant scraper sending blank Content-Type and no fallback_scrape_protocol specified for target")
```
Comment (beorn7) [inline model/textparse/interface.go:90]: Don't capitalize the error message. See https://go.dev/wiki/CodeReviewComments#error-strings (and other error messages in this package).

File: model/textparse/interface.go [hunk 5]
```diff
@@ -78,28 +80,65 @@ type Parser interface {
...
+		}
+		return fallbackType, errors.New("Non-compliant scraper sending blank Content-Type, using fallback_scrape_protocol for target")
```
Comment (beorn7) [inline model/textparse/interface.go:92]: Same.

File: model/textparse/interface.go [hunk 6]
```diff
@@ -78,28 +80,65 @@ type Parser interface {
...
 	}
 
+	// We have a contentType, parse it.
 	mediaType, _, err := mime.ParseMediaType(contentType)
 	if err != nil {
-		return NewPromParser(b, st), err
+		if fallbackType == "" {
+			retErr := errors.New("Cannot parse Content-Type and no fallback_scrape_protocol for target")
```
Comment (beorn7) [inline model/textparse/interface.go:99]: Same.

File: model/textparse/interface.go [hunk 7]
```diff
@@ -78,28 +80,65 @@ type Parser interface {
...
+			return "", errors.Join(retErr, err)
+		}
+		retErr := errors.New("Could not parse received Content-Type, using fallback_scrape_protocol for target")
```
Comment (beorn7) [inline model/textparse/interface.go:102]: Same.

Maybe let's include the used fallbackType in the error message.

File: model/textparse/interface.go [hunk 8]
```diff
@@ -78,28 +80,65 @@ type Parser interface {
...
+
+	// We have a valid media type, either we recognise it and can use it
+	// or we have to error.
+	switch mediaType {
+	case "application/openmetrics-text", "application/vnd.google.protobuf", "text/plain":
+		return mediaType, nil
+	}
+	// We're here because we have no recognised mediaType.
+	if fallbackType == "" {
+		return "", fmt.Errorf("Scraper sending unrecognisable Content-Type '%q' and no fallback_scrape_protocol specified for target", contentType)
```
Comment (beorn7) [inline model/textparse/interface.go:114]: Sorry for the list of nits:
- Don't capitalize the error message, see above.
- The "scraper" is us. What we scrape is better called "target".
- `%q` already quotes the string, so we don't need additional quotes around it in the format string.

Comment (beorn7) [reply model/textparse/interface.go:114]: In different news: Should we maybe say "unsupported" rather than "unrecognisable"?

Comment (beorn7) [reply model/textparse/interface.go:114]: Maybe like this:

```suggestion
		return "", fmt.Errorf("received unsupported Content-Type %q and no fallback_scrape_protocol specified", contentType)
```

File: model/textparse/interface.go [hunk 9]
```diff
@@ -78,28 +80,65 @@ type Parser interface {
...
 	}
+	return fallbackType, fmt.Errorf("Content-Type '%q' not recognised mediaType, using fallback_scrape_protocol for target", contentType)
```
Comment (beorn7) [inline model/textparse/interface.go:116]: See above. Also, we are conflating mediaType and ContentType now. Maybe let's use exactly the same error message as above, just that now we should include the fallback picked.
Maybe like this:
```suggestion
	return fallbackType, fmt.Errorf("received unsupported Content-Type %q, using fallback_scrape_protocol %q", contentType, fallbackType)
```

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks good. Just final nitpicking about the error messages.

@bwplotka does this look good to you? If you have no objections, we can merge once my nits have been addressed.

File: model/textparse/interface.go (hunk 4 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Fixed.

File: model/textparse/interface.go (hunk 5 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Fixed.

File: model/textparse/interface.go (hunk 6 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Fixed.

File: model/textparse/interface.go (hunk 7 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Done!

File: model/textparse/interface.go (hunk 8 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Done!

File: model/textparse/interface.go (hunk 9 above)
Comment (alexgreenbank) [reply model/textparse/interface.go]: Done!

**Comment by alexgreenbank** [comment]
> @bwplotka does this look good to you? If you have no objections, we can merge once my nits have been addressed.

OK, all nits addressed, tests rejigged, merged conflicts fixed and all should be good. Almost there!

**Review by beorn7 — APPROVED** [review]
With no complaint received no far, let's :ship: .

**Comment by beorn7** [comment]
Note that this is a fallback for targets that are buggy (i.e. they do not correctly implement content negotiation). We intentionally removed the global default because it turned out to not be safe. Introducing a global option would partially defeat the purpose.

If so many of your targets are buggy that you need a global option, I recommend to prioritize fixing the targets.

**Pull request merged**